app/src/lib/stores/app-store.ts

```diff
@@ -2843,76 +2847,88 @@ export class AppStore extends TypedBaseStore<IAppState> {
     const { changesState, branchesState } = this.repositoryStateCache.get(
       repository
     )
-    const { tip } = branchesState
-    const currentBranch = tip.kind === TipState.Valid ? tip.branch : null
-    const hasChanges = changesState.workingDirectory.files.length > 0
-
-    if (
-      enableStashing() &&
-      hasChanges &&
-      currentBranch !== null &&
-      uncommittedChangesStrategy ===
-        UncommittedChangesStrategy.askForConfirmation
-    ) {
-      this._showPopup({
-        type: PopupType.StashAndSwitchBranch,
-        branchToCheckout: foundBranch,
-        repository,
-      })
 
-      return repository
-    }
+    let createdStashEntry = null
+    if (enableStashing()) {
+      const hasChanges = changesState.workingDirectory.files.length > 0
+      if (
+        hasChanges &&
+        uncommittedChangesStrategy.kind ===
+          UncommittedChangesStrategyKind.askForConfirmation
+      ) {
+        this._showPopup({
+          type: PopupType.StashAndSwitchBranch,
+          branchToCheckout: foundBranch,
+          repository,
+        })
 
-    let shouldPopStash = false
+        return repository
+      }
 
-    if (enableStashing() && currentBranch !== null) {
+      const { tip } = branchesState
+      const currentBranch = tip.kind === TipState.Valid ? tip.branch : null
       if (
-        uncommittedChangesStrategy ===
-        UncommittedChangesStrategy.stashOnCurrentBranch
+        currentBranch !== null &&
+        uncommittedChangesStrategy.kind ===
+          UncommittedChangesStrategyKind.stashOnCurrentBranch
       ) {
-        await this.createStash(repository, currentBranch.name)
+        await this._createStashAndDropPreviousEntry(
+          repository,
+          currentBranch.name
+        )
       } else if (
-        uncommittedChangesStrategy ===
-        UncommittedChangesStrategy.moveToNewBranch
+        uncommittedChangesStrategy.kind ===
+        UncommittedChangesStrategyKind.moveToNewBranch
       ) {
-        await createDesktopStashEntry(repository, foundBranch.name)
-        shouldPopStash = true
+        const hasDeletedFiles = changesState.workingDirectory.files.some(
+          file => file.status.kind === AppFileStatusKind.Deleted
+        )
+        if (hasDeletedFiles) {
+          await gitStore.performFailableOperation(() => {
+            return createDesktopStashEntry(repository, foundBranch.name)
+          })
```

Comment:
So it seems to me like we're gonna go ahead and load the last desktop stash entry regardless of whether or not the `createDesktopStashEntry` succeeded or not here given that we're wrapping the `createDesktopStashEntry` call in `performFailableOperation` but not verifying whether it succeeded or not which seems like it could lead to us popping the wrong stash entry.

Reply:
Yeah, I also think it'd be great here to have the result of `createDesktopStashEntry` set to `createdStashEntry`, rather than having this as two separate steps.

Reply:
b1cbf11 adds a check to ensure this was successful. 

> Yeah, I also think it'd be great here to have the result of createDesktopStashEntry set to createdStashEntry, rather than having this as two separate steps.

I like this idea. Gonna open an issue to track this so it doesn't get lost after we release.

Reply:
to add some more context here, `createDesktopStashEntry` on `development` returns `void`, not information about the stash created (or if it was successful). @iAmWillShepherd has done a minimal fix in b1cbf11 (by returning `true` on success) that should address the case @niik called out and prevent us from accidentally popping the wrong stash (or trying to pop nothing).

i agree with @iAmWillShepherd that changing `createDesktopStashEntry` to return a `sha` or `stashEntry` on success is a good idea that we can address in a subsequent PR, since that requires some result parsing and validation.